x/restake/keeper/msg_server_test.go

```diff
@@ -0,0 +1,279 @@
+			expErrMsg: "not allowed denom",
+			preCheck:  func() {},
+			postCheck: func() {},
+		},
+		{
+			name: "mix both allow and unallow denom",
+			input: &types.MsgStake{
+				StakerAddress: ValidAddress1.String(),
+				Coins: sdk.NewCoins(
+					sdk.NewCoin("usdt", sdkmath.NewInt(10)),
+					sdk.NewCoin("uband", sdkmath.NewInt(10)),
+				),
+			},
+			expErr:    true,
+			expErrMsg: "not allowed denom",
+			preCheck:  func() {},
+			postCheck: func() {},
+		},
+		{
+			name: "success - have previous stake",
+			input: &types.MsgStake{
+				StakerAddress: ValidAddress1.String(),
+				Coins: sdk.NewCoins(
+					sdk.NewCoin("uband", sdkmath.NewInt(10)),
+				),
+			},
+			expErr:    false,
+			expErrMsg: "",
+			preCheck: func() {
+				suite.bankKeeper.EXPECT().
+					SendCoinsFromAccountToModule(gomock.Any(), ValidAddress1, types.ModuleName, sdk.NewCoins(
+						sdk.NewCoin("uband", sdkmath.NewInt(10)),
+					)).
+					Return(nil).
+					Times(1)
+			},
+			postCheck: func() {
+				stake := suite.restakeKeeper.GetStake(ctx, ValidAddress1)
+				suite.Require().Equal(types.Stake{
+					StakerAddress: ValidAddress1.String(),
+					Coins: sdk.NewCoins(
+						sdk.NewCoin("uband", sdkmath.NewInt(60)),
+					),
+				}, stake)
+			},
+		},
+		{
+			name: "success - no previous stake",
+			input: &types.MsgStake{
+				StakerAddress: ValidAddress2.String(),
+				Coins: sdk.NewCoins(
+					sdk.NewCoin("uband", sdkmath.NewInt(10)),
+				),
+			},
+			expErr:    false,
+			expErrMsg: "",
+			preCheck: func() {
+				suite.bankKeeper.EXPECT().
+					SendCoinsFromAccountToModule(gomock.Any(), ValidAddress2, types.ModuleName, sdk.NewCoins(
+						sdk.NewCoin("uband", sdkmath.NewInt(10)),
+					)).
+					Return(nil).
+					Times(1)
+			},
+			postCheck: func() {
+				stake := suite.restakeKeeper.GetStake(ctx, ValidAddress2)
+				suite.Require().Equal(types.Stake{
+					StakerAddress: ValidAddress2.String(),
+					Coins: sdk.NewCoins(
+						sdk.NewCoin("uband", sdkmath.NewInt(10)),
+					),
+				}, stake)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			tc.preCheck()
+			_, err := suite.msgServer.Stake(suite.ctx, tc.input)
+
+			if tc.expErr {
+				suite.Require().Error(err)
+				suite.Require().Contains(err.Error(), tc.expErrMsg)
+			} else {
+				suite.Require().NoError(err)
+			}
+
+			tc.postCheck()
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestMsgUnstake() {
+	ctx := suite.ctx
+	suite.setupState()
+
+	testCases := []struct {
+		name      string
+		input     *types.MsgUnstake
+		expErr    bool
+		expErrMsg string
+		preCheck  func()
+		postCheck func()
+	}{
+		{
+			name: "unstake more than staked coins",
+			input: &types.MsgUnstake{
+				StakerAddress: ValidAddress1.String(),
+				Coins: sdk.NewCoins(
+					sdk.NewCoin("uband", sdkmath.NewInt(2000)),
+				),
+			},
+			expErr:    true,
+			expErrMsg: "stake not enough",
+			preCheck:  func() {},
+			postCheck: func() {},
+		},
+		{
+			name: "success",
```

Comment:
should we add case of not enough total power after unstake

Reply:
added